src/dosext/dpmi/dpmi.c

```diff
@@ -404,6 +409,18 @@ int dpmi_is_valid_range(dosaddr_t addr, int len)
   return 1;
 }
 
+int dpmi_read_access(dosaddr_t addr)
+{
+  dpmi_pm_block *blk = lookup_pm_blocks_by_addr(addr);
+  return blk && (blk->attrs[(addr - blk->base) >> PAGE_SHIFT] & 7) > 0;
+}
+
+int dpmi_write_access(dosaddr_t addr)
+{
+  dpmi_pm_block *blk = lookup_pm_blocks_by_addr(addr);
+  return blk && (blk->attrs[(addr - blk->base) >> PAGE_SHIFT] & 0xf) > 8;
```

Comment:
Maybe use bitmasks instead of `>`?

Reply:
I'm concerned about the value "8" here (R/W but uncommitted, so PROT_NONE). There's this disabled section in `memory.c`:
```
#if 0
        /* DPMI spec says that RW bit can only be changed on committed
         * page, but some apps (Elite First Encounter) is changing it
         * also on uncommitted. */
        if (!com && !old_com) {
          D_printf(" Not changing RW(-) on uncommitted page\n");
          return 0;
        }
#endif
```
which implies that sometimes attrs are set to 8.

Reply:
Yes, but writable still means `attr&9 == 9`, no?

Reply:
No, that won't catch the case attr==0xa (writable shared memory). The two valid values I've seen for `attr&0xf` are `0x9` and `0xa`.

Reply:
But shared had to be 0xb! :)
So we can either change shared to 3,
or just leave things as you did.
Whatever you prefer then.

Reply:
The table in the DPMI spec says 2, so.. (http://www.delorie.com/djgpp/doc/dpmi/api/310506.html)

Reply:
Well, it says "mapped". :)
Whatever it is. File-mapped likely.
SHM can well be taken from "unused" space.
SHM is really something that only DPMIONE
implemented, so we are free to do whatever
we like. But its up to you.

Reply:
Hmm when reading the DPMI spec it should actually be 1, not 2 or 3. And 2 is only used with 0x508 (device mapping) which we don't implement.

Unless DPMIONE uses 2?

In practise I care about the value since it's used with the LDT in msdos_ldt.c, which uses
```
    for (i = 0; i < npages; i++)
        attrs[i] = 0x83;        // NX, RO
    DPMISetPageAttributes(shm.handle, 0, attrs, npages);
```
but DPMISetPageAttributes actually sets 0x82, since it was set to 0xa initially and then then 3 in 0x83 means "modify attributes without changing page type".

Reply:
> Hmm when reading the DPMI spec it should actually be 1, not 2 or 3.

Where is that said?

> And 2 is only used with 0x508 (device mapping) which we don't implement.

We probably can use it for hwram too,
which I've added now?

> but DPMISetPageAttributes actually sets 0x82

Yes, that's right.
Can still be 7 then, but as it is explicitly
not allowed, then some bit from 7..15 area,
where I also put NX.

Reply:
In fact, bit 2 can be used.
7 is not allowed when setting an
attribute, but you never set it by hands.
You'll still do 0x83. But GetPageAttr
will return 0x85.

Reply:
There are 2 kind of bits.
Some you can modify (RW, NX), some
not (mapped, SHM). Those that are not,
are _forbidden_ in SetPageAttr. Which
means, SHM should be there too.

Reply:
DPMI says on http://www.delorie.com/djgpp/doc/dpmi/api/310506.html that "Mapped pages can only occur in memory blocks under DPMI hosts that support the Device Mapping capability or the Conventional Memory Mapping capability. See Int 31H Functions 0401H, 0508H, and 0509H."

Indeed on SetPageAttr only for bits 0-2 only 0,1,3 are allowed, and for GetPageAttr, only 
0-2 are defined, so bit 2 can be set internally, but will need to be masked out in the API.
So if we use bit 2 internally to denote "shared" then shared memory would use 8+4+1=0xd by default, 0x80+4+1=0x85 for the LDT, and GetPageAttr (0x0506) would return 0x81 for the LDT, masking out ~0x04.

Reply:
It doesn't need to be masked out in API.
It will be masked in SetPageAttr anyway
as it can't be changed, but GetPageAttr
should reveal it. "currently unused" doesn't
mean "should be zero" or alike. Its just
something to take from, and use.

Reply:
ok let's do 0x4|0x1 for committed shared then.

Reply:
Done.